js/kunai/ui/content.js

```diff
@@ -7,6 +7,9 @@ class Content {
     this.log.debug('initialzing...')
 
     this.log.debug(`found ${Badge.sanitize($('main[role="main"] div[itemtype="http://schema.org/Article"] .content-body span.cpp'))} badges`)
+
+    // 横幅を超える画像を横スクロール可能にするためにスクロール用のdivで囲む
+    $('div[itemprop="articleBody"]').find('img').wrap('<div class="scrollable">')
```

Comment:
これって最初からセレクターでimgタグまで絞り込んだほうが高速になったりしないですかね？

```suggestion
    $('div[itemprop="articleBody"] img').wrap('<div class="scrollable">')
```

Reply:
最近の JavaScript 事情はよく知らないですけれど、jQuery ではなくネイティブの `document.querySelectorAll` や `document.evaluate` を使った方が早いのでは。jQuery のソースを検索してみました ([querySelectorAll](https://github.com/jquery/jquery/search?q=querySelectorAll), [querySelector](https://github.com/jquery/jquery/search?q=querySelector), [evaluate](https://github.com/jquery/jquery/search?q=evaluate)) が、別に最近のブラウザに合わせて querySelector などを使った実装にしている訳でもない (?) みたいですし。後 jQuery 3.2.1 が使われているみたいですけれど、もう5年前のバージョンになるみたいですね。

Reply:
kunaiはユーザー側のブラウザ上で動きます
jqueryのコードは後で調査してみますね

Reply:
> kunaiはユーザー側のブラウザ上で動きます

なるほど。ありがとうございます。だとすると、疑問が…。この PR による DOM 修正は動的に行う必要のあるものなのでしょうか。静的に HTML を生成する段階で修正するべきの気がします。。

Reply:
https://github.com/cpprefjp/kunai/pull/119#discussion_r781957234

はい、静的に生成できるのであればそちらが好ましいと思います。

markdownをhtmlに変換している処理は
https://github.com/cpprefjp/site_generator/blob/master/run.py#L82-L92
あたりになると思うのですが、どのように設定すればimgをdivでラップできるのかわかりませんでした…
自分だと時間がかかるか詰まってしまうかしてしまいそうなため、どなたかにお願いする形になってしまいそうです…

Reply:
うーん。なるほど。状況が分かってきました。変換は外部ライブラリの [Python-Markdown/markdown](https://github.com/Python-Markdown/markdown) を使って行っているけれども、HTML 生成に対して何処まで外から設定を変更できるかは不明、ということですね。
- 実際に `img` 要素を生成しているのは [markdown/inlinepatterns.py:760](https://github.com/Python-Markdown/markdown/blob/9aa4586d92181a07d3c054d81ada9b100c879d68/markdown/inlinepatterns.py#L760-L777) のようですけれども、外部設定でここの振る舞いを直接変更できそうな気配はありません。
- これの呼び出し元 [markdown/treeprocessors.py:272](https://github.com/Python-Markdown/markdown/blob/a11431539d08e14b0bd821ceb101fa59d6a74c8a/markdown/treeprocessors.py#L272) を見ると、生成された `node` は `__stashNode` を通じて [`self.stashed_nodes`](https://github.com/Python-Markdown/markdown/blob/a11431539d08e14b0bd821ceb101fa59d6a74c8a/markdown/treeprocessors.py#L111) 配列に一旦待避された後に、全て文字列を変換した後<sup>[推測]</sup>に [また取り出されて](https://github.com/Python-Markdown/markdown/blob/a11431539d08e14b0bd821ceb101fa59d6a74c8a/markdown/treeprocessors.py#L207) 最終的に [DOM Tree を表す木構造の中に埋め込まれる](https://github.com/Python-Markdown/markdown/blob/a11431539d08e14b0bd821ceb101fa59d6a74c8a/markdown/treeprocessors.py#L356-L398)。
- うーん。拡張 (extension) として `Treeprocessor` を実装して登録すれば行けるのではないかと予想。`Treeprocessor` の extension の実装例は [Search · treeprocessors.register](https://github.com/Python-Markdown/markdown/search?q=treeprocessors.register) などに。具体的には `img` を `div>img` に置き換える extension を作って [cpprefjp/markdown_to_html](https://github.com/cpprefjp/markdown_to_html) に入れて、それを [例の箇所から参照](https://github.com/cpprefjp/site_generator/blob/master/run.py#L82-L92) したら行ける…かも。

Reply:
普通に吐かれたmarkdownに対してもう1段階の変換器を挟むほうがmarkdown生成器への依存が減っていいのではという気がしています。具体的にはjqueryをnode.jsで動かして上のような変換するとか・・・。

Reply:
とはいえすでにいろいろ拡張しているから今更という話もあるか・・・

Reply:
PyhtonレイヤーとJSレイヤーどちらにしましょう…

Reply:
Pyhtonレイヤーの調査チケットを切ってこちらはpending...?

Reply:
> 普通に吐かれたmarkdownに対してもう1段階の変換器を挟むほうがmarkdown生成器への依存が減っていいのではという気がしています。具体的にはjqueryをnode.jsで動かして上のような変換するとか・・・。

JavaScript で処理したいなってのは本当に思いました。HTML を取り扱うのであれば Python じゃなくて JavaScript なのだなぁと実感しました。やっぱり言語には、既存のライブラリ・API的に、向き不向きって物凄くあるのだなぁと (Python の markdown ライブラリを作った人は、何故 Python で Markdown to HTML 変換器を書いてしまったのか)。

今は別に Python の変換後に Node の処理を走らせているという訳ではない (?) のですよね。だとするとこれだけの為に Node で parse して処理して再出力するという処理を加えるのも、細かい調整が色々面倒そうだし (例えば parse してそのまま再出力した時に全く同じになるのかとかのチェックとか) 大袈裟だなという気はします。一方で、一旦 JavaScript を使う仕組を作ってしまえば、今後似たようなことがあった時にやりやすいというか選択肢・可能性が拡がるような気もします。

作業に入る前に一旦、元の枠組みを作って下さった方々の意向も確認して良いような気がします。頑張って作った後にやっぱり駄目ってなったら悲しいので。何れにしても CSS は記述しなければならないので、この PR は仰る通り取り敢えずは開いたまま pending ですかね。

Reply:
memo: extension作るならこの辺使えそう・・・？https://github.com/Python-Markdown/markdown/blob/e11cd255cae5fd3c5ef5fdd6352cd28e212fd328/markdown/extensions/codehilite.py#L235-L252

Reply:
この例は (呼び出している外部クラス・関数などを除けば) 短くて良さそうですね

Reply:
@melpon さん @faithandbrave さんの意向が未だわからないですけれども、まあ取り敢えず Python で作って撥ねられることはない気がするので、気が向けば試しに作ってしまって良いかもと思います。

Reply:
Markdown -> HTMLの変換ライブラリも依存しすぎるとリプレースが大変なのは確かなので、変換後のHTMLにさらに変換をかけるレイヤーはあってもいい気がしますね。今後がラクになるので。

Reply:
このPRはそんなに大きくないですし、困っているようであればとりあえずマージしてしまって、別な仕組みを入れる際にrevertするのでもよいと思います。

Reply:
画像はそれほど使わないので速度はそこまで重要ではない気がしますが、 @nekko1119 さんどうでしょう。

Reply:
長らく放置されてしまいましたが、マージに向けて動きたいです。
私の当初指摘はマージするための阻害要因にはならない認識です。
@faithandbrave